Make the multi-threading interface a bit more flexible

dmhacker opened the pull request:
Tackles #43 .

A lot of the earlier multi-threading code had a lot of cruft that wasn't really doing much. For example: the unnecessary spawning of an entire scheduled executor service per scraper per company, when that service had only 1 thread itself and the main thread per company was just blocking on the individual service. 

The new proposal is this: keep a massive scheduled executor service available to all scrapers. Now, treat each scraper process as simply a chain of Runnable's (maybe with delay) to be executed in sequence and ending a callback that returns a completed list of positions. The naive BFSStrategy will simply be a giant Runnable that does the entire BFS process + scraping + waiting without deferring to the service. The faster BFSMTStrategy will be better broken up as such: 

1. Find current candidate & fetch the page. EXECUTE IMMEDIATELY.
2. Process page, extract position if there is one, go back to 1. EXECUTE AFTER PAGE LOAD DELAY.
3. If no more candidates exist after 1 ends, execute callback.

This seemingly simple change yields much faster results. Why? BFSStrategy runs all in one thread and to completion. This means that if you add 3N BFSStrategies to the executor service with N threads, the executor will complete the first N strategies added, then the second N strategies added, then the third N strategies added. When the N threads are stuck waiting, they are unable to do anything else. However, since BFSMTStrategy is broken up into separate tasks, this means that rather than waiting for a page to load, a thread can simply drop its task back into the pool and go try to take another available task. In this manner, the BFSMTStrategy runnables are continuously passed between threads, leading to a performance increase.  (Note that there is still a performance increase using BFSStrategy + the scheduled executor service, which is what is enabled right now - more on this below). 

However, BFSMTStrategy is 'broken' at the moment and has been deprecated. This is because it waits after it releases control of the driver rather than during. Waiting during acquisition of a driver from the driver pool must handled entirely differently! In fact, it might be impossible to use the idea of a Runnable chain with how the driver pool is currently set up. This is because if a task holds on to a driver from the pool and is then put back into the service with a delay, it will get stuck behind other tasks waiting for the semaphore to be released. Essentially, it leads to deadlock. A solution might be to initialize the executor service with less threads than open drivers - this will be looked into later. Another problem is of course that you can't use a try-with block when transferring control of a driver acquisition to another task, which could lead to deadlock if either task fails and doesn't release the driver.

It should be pretty easy to switch out strategies, since all you have to do is change the one line in Main::scrapePositions() where they are initialized.

dmhacker committed:
Make stuff faster & remove individual executors, simplify multi-threading interface, deprecate BFSMT scraper

dmhacker commented:
Oh, and I should also add that obviously the main speed decrease is going to be the web driver pool. Even using BFSMTStrategy with a large number (24+) of threads, most of the threads are eventually going to end up stuck waiting for another thread to release a driver back into the pool. And for BFSStrategy, since tasks hold on to their web drivers when loading, other threads will also block waiting for them.

CTrando commented on src/main/java/com/internhub/data/Main.java:
```diff
@@ -114,15 +103,17 @@ private static void scrapePositions(String name) {
+                positionScraper.fetch(company, positions::addAll);
             }
         }
+        PositionWriter writer = new PositionHibernateWriter();
```
We should write as we scrape so if the scraper crashes we don't lose all progress.

CTrando commented on src/main/java/com/internhub/data/positions/scrapers/PositionCallback.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.List;
+
+public interface PositionCallback {
```
maybe just use a runnable here, it's so similar anyways

CTrando replied:
also we are using the convention `I-` for interfaces. Look into anonymous classes too I think this is unnecessary

dmhacker replied:
I was thinking about using a runnable, but it would still require us to extend the class. The idea behind the callback class was just for it to run with the finished results upon completion, so we don't have to worry resolving the function stack with return arguments. Will look more into this.

CTrando commented on src/main/java/com/internhub/data/positions/scrapers/PositionScraper.java:
```diff
@@ -1,22 +1,32 @@
+    }
+
+    public void setup(Company company) {
+        mLinkCache.put(company.getName(), mInitialStrategy.fetchInitialLinks(company));
```
what's this second call doing here? Does it have some side effect you are using? If it does then it's bad practice cause it doesn't make sense while reading it

dmhacker replied:
Yeah, that was mainly used so that we can treat it as a separate task that the scheduled executor service can process concurrently along with the actual scraping. See Main::scrapePositions() to see how it's used. I admit though, it's not the best design.

CTrando commented on src/main/java/com/internhub/data/positions/scrapers/strategies/IPositionScraperStrategy.java:
```diff
@@ -10,5 +10,5 @@
 public interface IPositionScraperStrategy {
     Logger logger = LoggerFactory.getLogger(PositionScraper.class);
-    List<Position> fetch(Company company, List<String> initialLinks);
+    void fetch(Company company, List<String> initialLinks, PositionCallback callback);
```
consider overloading the method so one doesn't require the callback

CTrando commented on src/main/java/com/internhub/data/positions/scrapers/strategies/impl/GoogleInitialLinkStrategy.java:
```diff
@@ -31,8 +31,11 @@ public GoogleInitialLinkStrategy() {
-        if(googled.isEmpty()) {
-            logger.warn("Link strategy could not find any initial links.");
+        if (googled.isEmpty()) {
+            logger.warn(String.format("Link strategy could not find any initial links for %s.", company.getName()));
```
maybe add some quotes around the company name

CTrando commented on src/main/java/com/internhub/data/positions/scrapers/strategies/impl/PositionBFSStrategy.java:
```diff
@@ -41,55 +44,57 @@ public PositionBFSStrategy(WebDriver driver) {
+            try {
+                Thread.sleep(PAGE_LOAD_DELAY_MS);
+            } catch (InterruptedException e) {
+                logger.error("Could not wait for page to load.", e);
```
Could not wait for javascript to load*

CTrando commented on src/main/java/com/internhub/data/Main.java:
```diff
@@ -66,40 +62,33 @@ private static void scrapeCompanies() {
-        List<Callable<List<Position>>> tasks = Lists.newArrayList();
-        try (MyWebDriverPool pool = new MyWebDriverPool()) {
-            InitialLinkStrategy linkStrategy = new GoogleInitialLinkStrategy();
+        ScheduledExecutorService executor = Executors.newScheduledThreadPool(24);
```
is 24 a safe value?

dmhacker replied:
Eventually, the executor will be throttled by the amount of web drivers available, so it doesn't really matter what that is. As long as it isn't an insanely high number, I think it's ok.

CTrando commented on src/main/java/com/internhub/data/Main.java:
```diff
@@ -66,40 +62,33 @@ private static void scrapeCompanies() {
-        List<Future<List<Position>>> futures = Lists.newArrayList();
         try {
-            futures = executor.invokeAll(tasks);
+            latch.await();
```
shutdown the executor too

CTrando commented on src/main/java/com/internhub/data/Main.java:
```diff
@@ -66,40 +62,33 @@ private static void scrapeCompanies() {
+                executor.execute(() -> {
+                    positionScraper.setup(company);
+                    executor.execute(() -> {
+                        positionScraper.fetch(company, (intermediate) -> {
```
rename intermediate -> results

CTrando commented on src/main/java/com/internhub/data/Main.java:
```diff
@@ -114,15 +103,17 @@ private static void scrapePositions(String name) {
     }
 
     private static void scrapePositions(List<Company> companies) {
```
make it clearer that this method is for single threading. Why are we still maintaining single threading as well?

dmhacker replied:
Not sure. It should probably be transitioned over to the multi threading variant.

CTrando commented:
>  This is because if a task holds on to a driver from the pool and is then put back into the service with a delay, it will get stuck behind other tasks waiting for the semaphore to be released. Essentially, it leads to deadlock. 

How will this occur? How can a task hold on to a driver from the pool after it finishes executing?

CTrando commented:
> The new proposal is this: keep a massive scheduled executor service available to all scrapers. Now, treat each scraper process as simply a chain of Runnable's (maybe with delay) to be executed in sequence and ending a callback that returns a completed list of positions. The naive BFSStrategy will simply be a giant Runnable that does the entire BFS process + scraping + waiting without deferring to the service. 

We discussed this before and I thought you said it the performance increase would be negligible. Was the performance increase good enough to warrant these changes?

CTrando commented:
If you don't mind, can you provide some benchmarks between master and this branch. Just a depth of 20-25 should be enough to demonstrate noticeable differences.

dmhacker committed:
Address some minor issues

dmhacker committed:
Merge branch 'master' into fix/PageLoadMT

dmhacker committed:
Make Dockerfile less verbose

dmhacker committed:
Improve interface for position scraper by using a consumer instead of the list callback

dmhacker committed:
Add logback.xml file to shorten logger names

dmhacker committed:
Do some reorganizing to make code clearer & add dry-run option

dmhacker committed:
Pass all user-supplied arguments to docker script

dmhacker committed:
Fix docker start script to run correct command

dmhacker committed:
Fix dry-run option so that it actually works

dmhacker committed:
Fix MT BFS strategy hanging issue; clean unused imports

dmhacker committed:
Fix issue with dockerized Java not splitting CLI arguments

dmhacker committed:
Fix comment on previous issue

dmhacker commented:
Benchmarks will be updated as they finish.

Companies used:
- Amazon
- Google
- Tesla
- Citadel LLC
- IBM
- Intel
- Redfin
- SurveyMonkey
- Uber
- Workday
- Yelp
- Microsoft

Constants:
- Run on Surface Book 13" with 8-core Intel Core i7-8650U CPU with docker enabled.
- Maximum of 6 browser instances, maximum of 12 threads (master uses work stealing pool instead).
- Firefox, IntelliJ running in background but not being actively used.

Commands used:
1. **master** branch - BFSMTStrategy
     - `scripts/start_docker.sh -p`
     - `scrapePositions()` modified directly to run list of companies
2. **fix/PageLoadMT** branch - BFSStrategy 
     - `scripts/start_docker.sh "-s 'Amazon,Google,Tesla,Citadel LLC,IBM,Intel,Redfin,SurveyMonkey,Uber,Workday,Yelp,Microsoft' -d"` 
    - Modified to use BFSStrategy internally
3. **fix/PageLoadMT** branch - BFSMTStrategy 
     - `scripts/start_docker.sh "-s 'Amazon,Google,Tesla,Citadel LLC,IBM,Intel,Redfin,SurveyMonkey,Uber,Workday,Yelp,Microsoft' -d"` 
    - Modified to use BFSMTStrategy internally

Note that Citadel LLC has a space; the command is not split up there.

Results (best of two trials):
1. 20 minutes, 4 seconds.
2. 24 minutes, 30 seconds.
3. 21 minutes, 25 seconds.

Conclusion:
Performance increases in terms of time are pretty negligible no matter which multi-threading design you choose. Only ways we could speed this up are 1) increase number of web drivers, 2) refractor actual scraper algorithm to run in parallel, 3) split the workload across multiple machines. These changes will reduce the number of active threads at any given point in time however.

dmhacker committed:
Add log rolling, store all logs in directory

dmhacker committed:
Bind-mount logs folder to Docker container

dmhacker committed:
Rename position scraper involving executor service

CTrando commented on .gitignore:
```diff
@@ -9,6 +9,7 @@
 
 # Log files
 *.log*
+/logs/
```
Do you need a starting forward slash here?

CTrando replied:
As in `logs/` instead of `/logs/`

CTrando commented on src/main/java/com/internhub/data/Main.java:
```diff
@@ -11,6 +11,7 @@
 import com.internhub.data.positions.scrapers.IPositionScraper;
 import com.internhub.data.positions.scrapers.ScheduledPositionScraper;
 import com.internhub.data.positions.scrapers.strategies.impl.GoogleInitialLinkStrategy;
+import com.internhub.data.positions.scrapers.strategies.impl.PositionBFSMTStrategy;
```
does this import get used?
